cpp/src/io/json/host_tree_algorithms.cu

```diff
@@ -805,4 +809,607 @@ void scatter_offsets(
+    if (is_enabled_mixed_types_as_string) {
+      return cudf::detail::make_std_vector_sync(
+        is_all_nulls_each_column(input, column_tree, tree, col_ids, options, stream), stream);
+    }
+    return std::vector<uint8_t>();
+  }();
+  auto [ignore_vals, columns] = build_tree2(root,
+                                            is_str_column_all_nulls,
+                                            d_column_tree,
+                                            d_unique_col_ids,
+                                            d_max_row_offsets,
+                                            column_names,
+                                            row_array_parent_col_id,
+                                            is_array_of_arrays,
+                                            options,
+                                            stream,
+                                            mr);
+  if (ignore_vals.empty()) return;
+  scatter_offsets(tree,
+                  col_ids,
+                  row_offsets,
+                  node_ids,
+                  sorted_col_ids,
+                  d_column_tree,
+                  ignore_vals,
+                  columns,
+                  stream);
+}
+
+std::pair<thrust::host_vector<uint8_t>,
+          std::unordered_map<NodeIndexT, std::reference_wrapper<device_json_column>>>
+build_tree2(device_json_column& root,
+            std::vector<uint8_t> const& is_str_column_all_nulls,
+            tree_meta_t& d_column_tree,
+            device_span<NodeIndexT const> d_unique_col_ids,
+            device_span<size_type const> d_max_row_offsets,
+            std::vector<std::string> const& column_names,
+            NodeIndexT row_array_parent_col_id,
+            bool is_array_of_arrays,
+            cudf::io::json_reader_options const& options,
+            rmm::cuda_stream_view stream,
+            rmm::device_async_resource_ref mr)
+{
+  CUDF_FUNC_RANGE();
+  bool const is_enabled_lines                 = options.is_enabled_lines();
+  bool const is_enabled_mixed_types_as_string = options.is_enabled_mixed_types_as_string();
+  auto unique_col_ids = cudf::detail::make_host_vector_async(d_unique_col_ids, stream);
+  auto column_categories =
+    cudf::detail::make_host_vector_async(d_column_tree.node_categories, stream);
+  auto const column_parent_ids =
+    cudf::detail::make_host_vector_async(d_column_tree.parent_node_ids, stream);
+  auto column_range_beg =
+    cudf::detail::make_host_vector_async(d_column_tree.node_range_begin, stream);
+  auto const max_row_offsets = cudf::detail::make_host_vector_async(d_max_row_offsets, stream);
+  auto num_columns           = d_unique_col_ids.size();
+
+  auto to_json_col_type = [](auto category) {
+    switch (category) {
+      case NC_STRUCT: return json_col_t::StructColumn;
+      case NC_LIST: return json_col_t::ListColumn;
+      case NC_STR: [[fallthrough]];
+      case NC_VAL: return json_col_t::StringColumn;
```

Comment:
did you mean to have NC_STR return StringColumn? then it should be  ```case NC_STR: case NC_VAL:```

Reply:
Yes. Both NC_STR and NC_VAL will return StringColumn. 
Even though the json tokenizer gives extra information based on wrapped quoted strings as NC_STR, it's not useful for parser at the moment.